### PR TITLE
Switch from Strings to values to keep the maximum information, better messages

### PR DESCRIPTION
**Pull request by ThibaudDauce**

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -1,22 +1,43 @@
...
-#[derive(Debug, Error, PartialEq)]
+#[derive(Debug, Error)]
 pub enum CryptoCoreError {
-    #[error("Wrong size: {given} given should be {expected}")]
-    SizeError { given: usize, expected: usize },
-    #[error("Invalid size: {0}")]
-    InvalidSize(String),
+    #[error("wrong size when parsing bytes: {given} given should be {expected}")]
+    DeserialisationSizeError { given: usize, expected: usize },
+    #[error("when reading LEB128 size, {0}")]
```
Comment (tbrezot) [inline src/error.rs:8]:
```suggestion
    #[error("when reading LEB128, {0}")]
```

File: src/error.rs [hunk 2]
```diff
@@ -1,22 +1,43 @@
...
+    ReadLeb128Error(leb128::read::Error),
+    #[error("deserialisation error: {0}")]
+    GenericDeserialisationError(String),
+
+    #[error("when writing {value} as LEB128 size, IO error {error}")]
+    WriteLeb128Error { value: u64, error: std::io::Error },
+
+    #[error("when writing {bytes_len} bytes, IO error {error}")]
```
Comment (tbrezot) [inline src/error.rs:16]:
```suggestion
    #[error("when writing {bytes_len} bytes, {error}")]
```

File: src/symmetric_crypto/aes_256_gcm_pure/mod.rs [hunk 3]
```diff
@@ -47,11 +47,10 @@ impl Dem<KEY_LENGTH> for Aes256GcmCrypto {
...
         if plaintext.len() as u64 > MAX_PLAINTEXT_LENGTH {
-            return Err(CryptoCoreError::InvalidSize(format!(
-                "Plaintext is too large ({} bytes), max size: {} ",
-                plaintext.len(),
-                MAX_PLAINTEXT_LENGTH
-            )));
+            return Err(CryptoCoreError::PlaintextTooBigError {
+                plaintext_len: plaintext.len(),
+                max: MAX_PLAINTEXT_LENGTH,
+            });
```
Comment (tbrezot) [inline src/symmetric_crypto/aes_256_gcm_pure/mod.rs:53]: I would not create a dedicated error, rather returning an `EncryptionError`.

File: src/symmetric_crypto/aes_256_gcm_pure/mod.rs [hunk 4]
```diff
@@ -72,18 +71,16 @@ impl Dem<KEY_LENGTH> for Aes256GcmCrypto {
...
         if ciphertext.len() < Self::ENCRYPTION_OVERHEAD {
-            return Err(CryptoCoreError::InvalidSize(format!(
-                "Ciphertext is too small ({} bytes), min size: {}",
-                ciphertext.len(),
-                Self::ENCRYPTION_OVERHEAD
-            )));
+            return Err(CryptoCoreError::CiphertextTooSmallError {
+                ciphertext_len: ciphertext.len(),
+                min: Self::ENCRYPTION_OVERHEAD as u64,
+            });
```
Comment (tbrezot) [inline src/symmetric_crypto/aes_256_gcm_pure/mod.rs:77]: I would not create a dedicated error, rather returning an DecryptionError.

File: src/symmetric_crypto/aes_256_gcm_pure/mod.rs [hunk 5]
```diff
@@ -72,18 +71,16 @@ impl Dem<KEY_LENGTH> for Aes256GcmCrypto {
...
         if ciphertext.len() as u64 > MAX_PLAINTEXT_LENGTH + Self::ENCRYPTION_OVERHEAD as u64 {
-            return Err(CryptoCoreError::InvalidSize(format!(
-                "Ciphertext is too large ({} bytes), max size: {} ",
-                ciphertext.len(),
-                MAX_PLAINTEXT_LENGTH + Self::ENCRYPTION_OVERHEAD as u64
-            )));
+            return Err(CryptoCoreError::CiphertextTooBigError {
+                ciphertext_len: ciphertext.len(),
+                max: MAX_PLAINTEXT_LENGTH + Self::ENCRYPTION_OVERHEAD as u64,
+            });
```
Comment (tbrezot) [inline src/symmetric_crypto/aes_256_gcm_pure/mod.rs:83]: Same here

File: src/symmetric_crypto/aes_256_gcm_pure/mod.rs (hunk 3 above)
Comment (ThibaudDauce) [reply src/symmetric_crypto/aes_256_gcm_pure/mod.rs:53]: Can this leak information? Because I think it's a really useful error for the user of the lib to known that what it's sent is too big…

Comment (ThibaudDauce) [reply src/symmetric_crypto/aes_256_gcm_pure/mod.rs:53]: I think we're leaking this information with the timing anyway… (since it's a early return)

Comment (tbrezot) [reply src/symmetric_crypto/aes_256_gcm_pure/mod.rs:53]: Indeed, a timing check could lead to this information. You can keep it like this if you prefer. Same for my other comments.

File: src/bytes_ser_de.rs [hunk 6]
```diff
@@ -34,19 +37,22 @@ pub trait Serializable: Sized {
     #[inline]
     fn try_from_bytes(bytes: &[u8]) -> Result<Self, Self::Error> {
         if bytes.is_empty() {
-            return Err(
-                CryptoCoreError::InvalidSize("Given byte string is empty".to_string()).into(),
-            );
+            return Err(CryptoCoreError::DeserialisationSizeError {
+                given: 0,
+                expected: size_of::<Self>(),
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:42]: No, this size is wrong because we may write LEB128 sizes in addition of the object's data. Because of the LEB128 format, you can only know a lower bound on the expected size (LEB128 sizes can range from 1 to several bytes).

File: src/bytes_ser_de.rs [hunk 7]
```diff
@@ -34,19 +37,22 @@ pub trait Serializable: Sized {
...
         if res.is_err() || de.finalize().is_empty() {
             res
         } else {
             // There should not be any more bytes to read
-            Err(CryptoCoreError::ConversionError(
-                "Remaining bytes after deserialization!".to_string(),
-            )
+            Err(CryptoCoreError::DeserialisationSizeError {
+                given: bytes.len(),
+                expected: size_of::<Self>(),
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:54]: Same here.

File: src/bytes_ser_de.rs [hunk 8]
```diff
@@ -68,22 +74,18 @@ impl<'a> Deserializer<'a> {
...
     pub fn read_array<const LENGTH: usize>(&mut self) -> Result<[u8; LENGTH], CryptoCoreError> {
         let mut buf = [0; LENGTH];
         self.readable.read_exact(&mut buf).map_err(|_| {
-            CryptoCoreError::InvalidSize(format!(
-                "Deserializer: failed reading array of: {LENGTH} bytes",
-            ))
+            CryptoCoreError::DeserialisationSizeError {
+                given: self.readable.len(),
+                expected: LENGTH,
+            }
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:88]: Are you sure an error on the size is the only possible kind of error? It is not clear to me when reading the doc of `read_exact()`:
```
If this function encounters an error of the kind
[`ErrorKind::Interrupted`](https://doc.rust-lang.org/nightly/std/io/error/enum.ErrorKind.html) then the error is ignored and the operation
will continue.

If this function encounters an "end of file" before completely filling
the buffer, it returns an error of the kind [`ErrorKind::UnexpectedEof`](https://doc.rust-lang.org/nightly/std/io/error/enum.ErrorKind.html).
The contents of `buf` are unspecified in this case.

If any other read error is encountered then this function immediately
returns. The contents of `buf` are unspecified in this case.

If this function returns an error, it is unspecified how many bytes it
has read, but it will never read more than would be necessary to
completely fill the buffer.
```

File: src/bytes_ser_de.rs [hunk 9]
```diff
@@ -99,17 +101,16 @@ impl<'a> Deserializer<'a> {
             return Ok(vec![]);
         };
         let len = usize::try_from(len_u64).map_err(|_| {
-            CryptoCoreError::InvalidSize(format!(
-                "Deserializer: size of array is too big: {} bytes",
-                len_u64
+            CryptoCoreError::GenericDeserialisationError(format!(
+                "size of array is too big for architecture: {len_u64} bytes",
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:105]:
```suggestion
                "size of vector is too big for architecture: {len_u64} bytes",
```

File: src/bytes_ser_de.rs [hunk 10]
```diff
@@ -99,17 +101,16 @@ impl<'a> Deserializer<'a> {
...
             ))
         })?;
         let mut buf = vec![0_u8; len];
         self.readable.read_exact(&mut buf).map_err(|_| {
-            CryptoCoreError::InvalidSize(format!(
-                "Deserializer: failed reading array of: {} bytes",
-                len
-            ))
+            CryptoCoreError::DeserialisationSizeError {
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:110]: Same here.

File: src/error.rs [hunk 11]
```diff
@@ -1,22 +1,43 @@
 use thiserror::Error;
 
 /// Error type for this crate.
-#[derive(Debug, Error, PartialEq)]
+#[derive(Debug, Error)]
```
Comment (tbrezot) [inline src/error.rs:4]: Why did you remove the `PartialEq` derivation ? Does this prevent from mapping on different cases ?

**Review by tbrezot — COMMENTED** [review]
Some last minute comments, sorry.

File: src/error.rs (hunk 11 above)
Comment (ThibaudDauce) [reply src/error.rs:4]: Because it was only used in the tests and I think some errors from `leb128` doesn't implement it.

File: src/bytes_ser_de.rs (hunk 8 above)
Comment (ThibaudDauce) [reply src/bytes_ser_de.rs:88]: `read_exact` has only one `Err` case from the code I think…

```
 #[inline]
    fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
        if buf.len() > self.len() {
            return Err(io::const_io_error!(
                ErrorKind::UnexpectedEof,
                "failed to fill whole buffer"
            ));
        }
        let (a, b) = self.split_at(buf.len());

        // First check if the amount of bytes we want to read is small:
        // `copy_from_slice` will generally expand to a call to `memcpy`, and
        // for a single byte the overhead is significant.
        if buf.len() == 1 {
            buf[0] = a[0];
        } else {
            buf.copy_from_slice(a);
        }

        *self = b;
        Ok(())
    }
```

File: src/bytes_ser_de.rs (hunk 6 above)
Comment (ThibaudDauce) [reply src/bytes_ser_de.rs:42]: Do we encode with LEB128 fixed sized arrays?

File: src/bytes_ser_de.rs [hunk 12]
```diff
@@ -37,23 +34,22 @@ pub trait Serializable: Sized {
...
+                } else {
+                    Err(CryptoCoreError::DeserialisationSizeError {
+                        given: bytes.len(),
+                        expected: result.length(),
+                    })?
+                }
             }
-            .into())
+            Err(err) => Err(err)?,
         }
```
Comment (tbrezot) [inline src/bytes_ser_de.rs:53]: You could also have unwrapped the error in early return, this seems to compile for me:
```
        let res = de.read();
        if res.is_err() || de.finalize().is_empty() {
            res
        } else {
            // There should not be any more bytes to read
            Err(CryptoCoreError::DeserialisationSizeError {
                given: bytes.len(),
                expected: res?.length(),
            }
            .into())
        }

```

**Review by tbrezot — APPROVED** [review]
